plugin.js

```diff
@@ -25,9 +25,9 @@ function getName (func) {
   return func.toString().split('\n').slice(0, 2).map(s => s.trim()).join(' -- ')
 }
 
-function Plugin (parent, func, opts, isAfter, timeout) {
+function Plugin (parent, func, optsOrFunc, isAfter, timeout) {
   this.func = func
-  this.opts = opts
+  this.opts = typeof optsOrFunc === 'function' ? optsOrFunc() : optsOrFunc
```

Comment:
```suggestion
  this.opts = typeof optsOrFunc === 'function' ? optsOrFunc(parent) : optsOrFunc
```

It might be helpful to call the opts function with the parent fastify instance. It will be handy so opts function could not rely on closures if they decide so.

Reply:
I agree, users might want to create local-scope functions. It is a simple change so I have added it along with the documentation in the latest commit.